java/ql/src/experimental/semmle/code/java/security/DecompressionBomb.qll

```diff
@@ -0,0 +1,713 @@
+   */
+  class InflateCall extends MethodAccess {
+    InflateCall() {
+      this.getReceiverType() instanceof TypeInflator and
+      this.getCallee().hasName("inflate")
+    }
+
+    /**
+     * A method Access as a sink which responsible for reading bytes
+     */
+    MethodAccess getAByteRead() { result = this }
+
+    // look at Zip4j comments for this method
+    predicate isControlledRead() { none() }
+  }
+
+  class InflateSink extends DecompressionBomb::Sink {
+    override predicate sink(DataFlow::Node sink, DataFlow::FlowState state) {
+      sink.asExpr() = any(InflateCall r).getAByteRead() and state = "inflator"
+    }
+  }
+
+  /**
+   * A type that is responsible for `ZipFile` Class
+   */
+  class TypeZipFile extends RefType {
+    TypeZipFile() { this.hasQualifiedName("java.util.zip", "ZipFile") }
+  }
+
+  /**
+   * Gets `n1` and `n2` which `ZipFile n2 = new ZipFile(n1);` or
+   * `InputStream n2 = zipFile.getInputStream(n1);` or `zipFile_As_n1.getInputStream(n2);`
+   */
+  private class ZipFileAdditionalTaintStep extends DecompressionBomb::AdditionalStep {
+    override predicate step(
+      DataFlow::Node n1, DataFlow::FlowState stateFrom, DataFlow::Node n2,
+      DataFlow::FlowState stateTo
+    ) {
+      (
+        exists(MethodAccess ma |
+          ma.getReceiverType() instanceof TypeZipFile and
+          ma = n2.asExpr() and
+          ma.getQualifier() = n1.asExpr() and
+          ma.getCallee().hasName("getInputStream")
+        )
+        or
+        exists(Call c |
+          c.getCallee().getDeclaringType() instanceof TypeZipFile and
+          c.getArgument(0) = n1.asExpr() and
+          c = n2.asExpr()
+        )
+      ) and
+      stateFrom = "ZipFile" and
+      stateTo = "ZipFile"
+    }
+  }
+}
+
+/**
+ * Providing InputStream and it subClasses that mostly related to Sinks of ZipFile Type,
+ * we can do
+ */
+module InputStream {
+  /**
+   * The Types that are responsible for `InputStream` Class and all classes that are child of InputStream Class
+   */
+  class TypeInputStream extends RefType {
+    TypeInputStream() { this.getASupertype*().hasQualifiedName("java.io", "InputStream") }
+  }
+
+  /**
+   * The methods that read bytes and belong to `InputStream` Type and all Types that are child of InputStream Type
+   */
+  class Read extends MethodAccess {
+    Read() {
+      this.getReceiverType() instanceof TypeInputStream and
+      this.getCallee().hasName(["read", "readNBytes", "readAllBytes"])
+    }
+  }
+
+  class ReadSink extends DecompressionBomb::Sink {
+    override predicate sink(DataFlow::Node sink, DataFlow::FlowState state) {
+      sink.asExpr() = any(Read r) and state = "ZipFile"
+    }
+  }
+
+  /**
+   * general additional taint steps for all inputStream and all Types that are child of inputStream
+   */
+  private class InputStreamAdditionalTaintStep extends DecompressionBomb::AdditionalStep {
+    override predicate step(
```

Comment:
The query flags any read on an InputStream as a "decompression" sink which is leading to thousands of FPs.

I think the problem is here since the taint step is applies to a read method on ANY class extending java.io.InputStream even those not related with decompression operations